qtfred/src/mission/dialogs/CampaignEditorDialogModel.h

```diff
@@ -0,0 +1,254 @@
+	inline void flagModified() { modified = true; }
+
+public: //clang
+private:
+	bool _saveTo(QString file) const;
+
+	template<typename T>
+	inline void modify(T &a, const T &b) {
+		if (a != b) {
+			a = b;
+			flagModified();
+		}
+	}
+	QStringList droppedMissions{};
+	bool modified{false};
+
+// Model state setters
+public slots:
+	inline void setCampaignName(const QString &name) {
+		modify<QString>(campaignName, name); }
+	inline void setCampaignTechReset(bool techReset) {
+		modify<bool>(campaignTechReset, techReset); }
+
+	void missionSelectionChanged(const QItemSelection &selected);
+
+	void setCurMnFirst();
+	inline void setCurMnBriefingCutscene(const QString &briefingCutscene) {
+		if (! mnData_it) return;
+		modify<QString>(mnData_it->mainhall, briefingCutscene); }
+	inline void setCurMnMainhall(const QString &mainhall) {
+		if (! mnData_it) return;
+		modify<QString>(mnData_it->mainhall, mainhall); }
+	inline void setCurMnDebriefingPersona(const QString &debriefingPersona) {
+		if (! mnData_it) return;
+		modify<QString>(mnData_it->debriefingPersona, debriefingPersona); }
+
+	int addCurMnBranchTo(const QModelIndex *other = nullptr, bool flip = false);
+	void delCurMnBranch(int node);
+
+	void selectCurBr(const QTreeWidgetItem *selected);
+	int setCurBrCond(const QString &sexp, const QString &mn, const QString &arg);
+	bool setCurBrSexp(int sexp);
+	void setCurBrIsLoop(bool isLoop);
+
+	void moveCurBr(bool up);
+
+	void setCurLoopAnim(const QString &anim);
+	void setCurLoopVoice(const QString &voice);
+
+//Model inner types
+private:
+	struct CampaignMissionData {
+		CampaignMissionData() = delete;
+		CampaignMissionData(QString file,
+							bool loaded = false,
+							const mission *fsoMn = nullptr,
+							const cmission *cm = nullptr);
+
+		static void initMissions(
+				const SCP_vector<SCP_string>::const_iterator &m_it,
+				CheckedDataListModel<CampaignEditorDialogModel::CampaignMissionData> &model);
+		void branchesFromFormula(CampaignEditorDialogModel *model, int formula, const cmission *loop = nullptr);
+
+	// state
+		const QString filename;
+
+		bool fredable{false};
+
+		const int nPlayers;
+		const QString notes;
+
+		const QStringList events;
+		const QStringList goals;
+
+		QString briefingCutscene;
+		QString mainhall;
+		QString debriefingPersona;
+
+		SCP_vector<CampaignBranchData> branches;
+	// state -- current branch
+		CampaignBranchData *brData_it{nullptr};
+		int brData_idx{-1};
+	};
+
+	struct CampaignBranchData {
+		explicit CampaignBranchData() = default;
+		CampaignBranchData(CampaignEditorDialogModel *model, int sexp_branch, const QString &from, const cmission *loop = nullptr);
+		CampaignBranchData(CampaignEditorDialogModel *model, const QString &from, QString to = "");
+
+		void connect(const SCP_unordered_set<const CampaignMissionData*>& missions);
+
+	// constants
+		enum BranchType { INVALID, REPEAT, NEXT, NEXT_NOT_FOUND, END, };
```

Comment:
Not having using-enum amenities, this would require sprinkling quite some amount of BranchType:: s throughout this header and the implementation file, reducing legibility.

Reply:
They're used by the implementation of branch validation/connection logic in CampaignEditorDialogModel.cpp, and private to that class? Otherwise, I'm sorry I don't understand your question.